src/controller/CHIPDeviceController.cpp

```diff
@@ -894,8 +894,34 @@ CHIP_ERROR DeviceCommissioner::StopPairing(NodeId remoteDeviceId)
 
 CHIP_ERROR DeviceCommissioner::UnpairDevice(NodeId remoteDeviceId)
 {
-    // TODO: Send unpairing message to the remote device.
-    return CHIP_NO_ERROR;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    MATTER_TRACE_EVENT_SCOPE("UnpairDevice", "DeviceCommissioner");
+
+    Platform::UniquePtr<DeviceUnpair> deviceUnpair(new DeviceUnpair());
+    deviceUnpair->RegisterCallback(this);
+    CommissioneeDeviceProxy * commissioneeDeviceProxy = nullptr;
+    if (CHIP_NO_ERROR == GetDeviceBeingCommissioned(remoteDeviceId, &commissioneeDeviceProxy))
+    {
+        deviceUnpair->UnpairDevice(commissioneeDeviceProxy, remoteDeviceId);
+    }
+    else
+    {
+        err = GetConnectedDevice(remoteDeviceId, &deviceUnpair->GetConnectedCallback(),
+                                 &deviceUnpair->GetConnectionFailureCallback());
+    }
+
+    mDeviceUnpair = std::move(deviceUnpair);
+
+    return err;
+}
+
+void DeviceCommissioner::OnDeviceUnpair(NodeId remoteDeviceId, CHIP_ERROR err)
+{
+    mDeviceUnpair = nullptr;
+    if (mPairingDelegate != nullptr)
+    {
+        mPairingDelegate->OnPairingDeleted(err);
```

Comment:
Which pairing?

Reply:
Delete PairingCallback interface was not used, so delete interface was used. Callback is needed to deliver the Unpair results, so if there is a better way, please suggest it.

Reply:
The point is, the existing signature of OnPairingDeleted is pretty broken, because it does not tell you which pairing.

I did suggest a better way: the process of removing a fabric from a node should just be encapsulated in a single object, instead of spread across the device controller and a helper object like this, and that object should notify its caller.  Similar to the way the commissioning window opener works.